lib/identity_cache/belongs_to_caching.rb

```diff
@@ -11,47 +11,17 @@ module ClassMethods
       def cache_belongs_to(association)
         ensure_base_model
 
-        unless association_reflection = reflect_on_association(association)
+        unless reflection = reflect_on_association(association)
           raise AssociationError, "Association named '#{association}' was not found on #{self.class}"
         end
-        if association_reflection.scope
+
+        if reflection.scope
           raise UnsupportedAssociationError, "caching association #{self}.#{association} is scoped which isn't supported"
         end
 
-        options = {}
-        self.cached_belongs_tos[association] = options
-
-        options[:embed]                   = false
-        options[:cached_accessor_name]    = "fetch_#{association}"
-        options[:records_variable_name]   = :"@cached_#{association}"
-        options[:association_reflection]  = association_reflection
-        options[:prepopulate_method_name] = "prepopulate_fetched_#{association}"
-
-        build_normalized_belongs_to_cache(association, options)
-      end
-
-      private
-
-      def build_normalized_belongs_to_cache(association, options)
-        foreign_key = options[:association_reflection].foreign_key
-        self.class_eval(<<-CODE, __FILE__, __LINE__ + 1)
-          def #{options[:cached_accessor_name]}
-            association_klass = association(:#{association}).klass
-            if association_klass.should_use_cache? && #{foreign_key}.present? && !association(:#{association}).loaded?
-              if instance_variable_defined?(:#{options[:records_variable_name]})
-                #{options[:records_variable_name]}
-              else
-                #{options[:records_variable_name]} = association_klass.fetch_by_id(#{foreign_key})
-              end
-            else
-              #{association}
-            end
-          end
+        cached_belongs_to = Cached::Association::Builder.build(:belongs_to, association, reflection: reflection)
```

Comment:
Ease of use. The builder's job is to translate the syntax we've been expecting in the configuration DSL to instantiate the right class. Cached associations are also not built at runtime so the extra cost associated with this indirection seems worth it to me. If we add more kinds of caching options in the future this seems more modular to me than `case`ing it out in the macro.

Reply:
I'm not worrying about performance cost.  The indirection causes a readability cost.  I don't see why it isn't the concern of the DSL method itself to translate the parameters to what we store internally.

> If we add more kinds of caching options in the future this seems more modular to me than `case`ing it out in the macro.

It looks like you are using a case statement for the other macros, so I don't see why it would be any different, other than moving the case statement further away from the more familiar API public method.

Reply:
Fair enough. I could just be modelling cached associations too closely to Active Record association builder pattern for no valid reason too. I'll change things to be more direct.